### ISSUE TITLE
[Question] Negative Sampling Methods for BPR

### ISSUE THREAD

**Issue by cocomoff** [RUCAIBox/RecBole#1391]
Hi Teams, thank you for your great RecBole library!

I'm now working on the BPR method on RecBole, and I appreciate if you give me any comments, links, and suggestions.

# What I want to implement
- I'm interested in how I can implement the method in the following paper and evaluate the result on RecBole.
  - [Jacek Wasilewski](https://dblp.org/pid/119/5812.html), [Neil Hurley](https://dblp.org/pid/86/2778.html): Bayesian Personalized Ranking for Novelty Enhancement. [UMAP 2019](https://dblp.org/db/conf/um/umap2019.html#WasilewskiH19): 144-148
- Since the idea basically follows the standard BPR, I think I can adopt [BPR implemented in RecBole](https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/general_recommender/bpr.py).
- The naive BPR trains its model on the data $D_S=\\{(u,i,j)\mid (u, i)\in\mathcal{R}, j\in \mathcal{I}\setminus\mathcal{I}_u^+\\}$, i.e., we need a tuple $(u,i,j)$ for user $u$, a positive item $i$, and negative item $i$.
- In the above paper, the authors suggested to prepare a dataset of $D_S^\mathrm{dist}=\\{(u,i,j)\mid (u, i)\in\mathcal{R}, j\in \mathcal{I}\setminus\mathcal{I}_u^+, \mathrm{dist}(i, j) < \tau\\}$, i.e., sample negative items with distance constraints.

# What I read and what I do not understand

- `Trainer` trains a BPR model; each learning step is seen in `_train_epoch` method.
  - Since I just uses the same loss function from the naive BPR, I do not need to customize here.
- `Trainer` prepares a dataset with an interface `TrainDataLoader` (`NegSampleDataLoader`)  and `Interaction` data type. 
  - By the default config setting for the BPR, the code [L.174-L.177 on abstract_dataloader.py](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/data/dataloader/abstract_dataloader.py#L174-L177) and `sampling_func` seems to prepare training data. 
  - Negative samples are generated in ``sample_by_user_ids` method. This function is from the `sampler` class.
  - Generated negative samples are formatted by `_neg_sample_by_pair_wise_sampling` method.
- `Sampler` has `self.sample_by_key_ids(user_ids, num)` on its base class `AbstractSampler` ([L.143-L179 here](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L143-L179)).  However, here, I got lost to follow what RecBole does for BPR (i.e., I cannot follow how negative samples are generated)
- In the document of [Customize Samplers](https://recbole.io/docs/developer_guide/customize_samplers.html), I am still not sure how I can design a customized negative samping methods.

I'm lookling forward to hearing from you if you have time. Thank you!

**Comment by leoleojie** [comment]
Hello, thanks for your attention to RecBole!
In BPR model, as you can see, the negative items are generated in `sample_by_key_ids`.
RecBole samples `num` negative items for each user in batch. We first implement  [`get_used_ids`](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L225-L246) to obtain `used_ids`, which are composed of positive items. When negative sampling, we will not sample from these items.  Then we implement [`self.sampling`](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L120-L134)to sample negative items. If the distribution is [`uniform`](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L222-L223), the negative items will be sampled by `np.random.randint` in a uniform distribution. Else if the distribution is [`popularity`](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L97-L118), the sample distribution will based on item popularity. After sampling, we will check whether these sampled items are in `used_ids`. If true, we will sample the reused items again until the sampled items do not duplicate the `used_ids`.

So if you want to sample negative items with distance constraints, you may need to rewrite the `self.sampling` by adding the constraints.

**Comment by cocomoff** [comment]
Hi @leoleojie, Thank you for your detailed comments. I'll try my implementation on [self.sampling](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L120-L134).

To check my understandings about samplers, I hope if you have time to see my another small question on negative sampling. I think I can understand `get_used_ids`. If we set `uniform` (default in config), the below code is what we talk about.

```python
def _uni_sampling(self, sample_num):
    return np.random.randint(1, self.item_num, sample_num)
```

Obviously, this just samples `sample_num` items of ranges [1, item_num]. However, I'm not sure that `sample_num` sampled items are _negative items_ since this sampling seems not to check any information of computed from `get_used_ids`. How can we know about this? I think I misunderstand some methods.

I'm lookling forward to hearing from you if you have time. Thank you!

**Comment by leoleojie** [comment]
@cocomoff 
In fact, we did not check for information about `get_used_ids` at `_uni_sampling`. However, as you can see in the [ while loop,](https://github.com/RUCAIBox/RecBole/blob/61848c0cc03d25645ebeffbfe9b58c48aaafaa76/recbole/sampler/sampler.py#L173-L179) after each sampling, we check the sampled items whether in `used_ids`. This might answer your question.

**Comment by cocomoff** [comment]
Hi @leoleojie, Thank you for your quick comment, and I understand what you mentioned in the while loop!

Again, thank you for your great RecBole library, and your support!

**Comment by leoleojie** [comment]
@cocomoff You're welcome
